File: app/src/main/java/com/zacharee1/systemuituner/LauncherActivity.java
package com.zacharee1.systemuituner;

import android.Manifest;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.crashlytics.android.Crashlytics;
import com.zacharee1.systemuituner.activites.info.IntroActivity;
import com.zacharee1.systemuituner.activites.instructions.SetupActivity;
import com.zacharee1.systemuituner.util.SuUtils;
import com.zacharee1.systemuituner.util.Utils;

import io.fabric.sdk.android.Fabric;

public class LauncherActivity extends AppCompatActivity
{

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        Fabric.with(this, new Crashlytics());

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);

        if (preferences.getBoolean("show_intro", true)) {
            startActivity(new Intent(this, IntroActivity.class));
        } else {
            String[] perms = new String[]{
                    Manifest.permission.WRITE_SECURE_SETTINGS,
                    Manifest.permission.DUMP,
                    Manifest.permission.PACKAGE_USAGE_STATS
            };

            String[] ret;

            if ((ret = Utils.checkPermissions(this, perms)).length > 0) {
                if (SuUtils.testSudo()) {
                    SuUtils.sudo("pm grant com.zacharee1.systemuituner android.permission.WRITE_SECURE_SETTINGS ; " +
                            "pm grant com.zacharee1.systemuituner android.permission.DUMP ; " +
                            "pm grant com.zacharee1.systemuituner android.permission.PACKAGE_USAGE_STATS");
                    Utils.startUp(this);
                    finish();
                } else {
                    Intent intent = new Intent(this, SetupActivity.class);
                    intent.putExtra("permission_needed", ret);
                    startActivity(intent);
                    finish();
                }
            } else {
                Utils.startUp(this);
                finish();
            }
        }

        finish();
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/LauncherActivity.kt
package com.zacharee1.systemuituner

import android.Manifest
import android.content.Intent
import android.os.Bundle
import androidx.preference.PreferenceManager
import com.crashlytics.android.Crashlytics
import com.zacharee1.systemuituner.activites.BaseAnimActivity
import com.zacharee1.systemuituner.activites.info.IntroActivity
import com.zacharee1.systemuituner.activites.instructions.SetupActivity
import com.zacharee1.systemuituner.util.SuUtils
import com.zacharee1.systemuituner.util.checkPermissions
import com.zacharee1.systemuituner.util.startUp
import io.fabric.sdk.android.Fabric

class LauncherActivity : BaseAnimActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        Fabric.with(this, Crashlytics())

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)

        if (preferences.getBoolean("show_intro", true)) {
            startActivity(Intent(this, IntroActivity::class.java))
        } else {
            val perms = arrayListOf(Manifest.permission.WRITE_SECURE_SETTINGS, Manifest.permission.DUMP, Manifest.permission.PACKAGE_USAGE_STATS)

            val ret = checkPermissions(perms)
            ret.removeAll(SetupActivity.NOT_REQUIRED)

            if (ret.isNotEmpty()) {
                if (SuUtils.testSudo()) {
                    SuUtils.sudo("pm grant $packageName ${Manifest.permission.WRITE_SECURE_SETTINGS} ; " +
                            "pm grant $packageName ${Manifest.permission.DUMP} ; " +
                            "pm grant $packageName ${Manifest.permission.PACKAGE_USAGE_STATS}")
                    startUp()
                    finish()
                } else {
                    SetupActivity.make(this, ret)
                    finish()
                }
            } else {
                startUp()
                finish()
            }
        }

        finish()
    }
}
